genesis/utils/geom.py

```diff
@@ -838,6 +838,77 @@ def transform_pos_quat_by_trans_quat(pos, quat, t_trans, t_quat):
     return new_pos, new_quat
 
 
+def transform_by_T(pos, T):
+    """
+    Transforms 3D points by a 4x4 transformation matrix or a batch of matrices,
+    supporting both NumPy arrays and PyTorch tensors.
+
+    Parameters
+    ----------
+    pos: np.ndarray | torch.Tensor
+        A numpy array or torch tensor of 3D points. Can be a single point
+         (3,), a batch of points (B, 3), or a batched batch of points (B, N, 3).
+    T: np.ndarray | torch.Tensor
+        The 4x4 transformation matrix or a batch of B transformation
+        matrices of shape (B, 4, 4). Must be of the same type as `pos`.
+
+    Returns
+    -------
+        The transformed points in a shape corresponding to the input dimensions.
+    """
+    assert pos.shape[-1] == 3, "Input positions must have 3 dimensions"
+
+    if isinstance(pos, torch.Tensor) and isinstance(T, torch.Tensor):
+        if T.ndim == 2:
+            T = T.unsqueeze(0)
+        if pos.ndim > 1:
+            ones_shape = pos.shape[:-1] + (1,)
+            pos_hom = torch.cat([pos, torch.ones(ones_shape, dtype=pos.dtype, device=pos.device)], dim=-1)
+        else:
+            pos_hom = torch.cat([pos, torch.tensor([1.0], dtype=pos.dtype, device=pos.device)])
+        if pos_hom.ndim == 1:
+            pos_hom = pos_hom.unsqueeze(0).unsqueeze(0)  # [1, 1, 4]
+        elif pos_hom.ndim == 2:
+            assert T.shape[0] == 1 or T.shape[0] == pos.shape[0]
+            pos_hom = pos_hom.unsqueeze(-2)  # [B, 1, 4]
+
+        pos_hom_t = pos_hom.swapaxes(-1, -2)  # (..., N, 4) -> (..., 4, N)
```

Comment:
Could you factorize your implementation to share more code between numpy and torch? If you favour `reshape` over `unsqueeze` and `squeeze` it should be possible to remove most of the duplication I think? That would be great !